File: src/patcher.cpp
#include <patcher.hpp>

#include <print>
#include <iostream>
#include <cstring>

using namespace noviy;

auto Patcher::build_iat_lookup_table() -> std::unordered_map<std::uint64_t, std::string> {
    std::unordered_map<std::uint64_t, std::string> iat_lookup_table;

    auto callback = [](void* context, const VA& addr, const std::string& module, const std::string& symbol) -> int {
        auto* table = static_cast<decltype(iat_lookup_table)*>(context);

        static constexpr std::array<const char*, 2> CD_CHECK_BYPASS = {
            "GetLogicalDriveStringsA",  // Lego Racers 2
            "GetLogicalDrives"          // Lego Rock Raiders / Lego Alpha Team
        };

        if (std::ranges::find(CD_CHECK_BYPASS, symbol) != CD_CHECK_BYPASS.end()) {
            std::print("Found {} in IAT: 0x{:x}\n", symbol, addr);
            (*table)[addr] = symbol;
        }

        return 0;
    };

    IterImpVAString(parser_.get(), callback, &iat_lookup_table);

    return iat_lookup_table;
}

auto Patcher::remove_relocation_entry(std::size_t offset_to_remove) {
    struct SectionInfo {
        std::uint64_t reloc_section_offset;
        std::uint64_t reloc_section_rva;
        std::uint32_t reloc_section_size;
    } section_info = {0, 0, 0};

    auto section_callback = [](void* cbd, const VA& base, const std::string& name, const image_section_header& sec,
                               const bounded_buffer* data) -> int {
        auto* info = static_cast<SectionInfo*>(cbd);

        if (name == ".reloc") {
            info->reloc_section_offset = sec.PointerToRawData;
            info->reloc_section_rva = sec.VirtualAddress;
            info->reloc_section_size = sec.SizeOfRawData;
            return 1;
        }

        return 0;
    };

    // iterate through sections to find the relocation section
    IterSec(parser_.get(), section_callback, &section_info);

    std::uint64_t reloc_section_offset = section_info.reloc_section_offset;
    std::uint64_t reloc_section_rva = section_info.reloc_section_rva;
    std::uint32_t reloc_section_size = section_info.reloc_section_size;

    if (reloc_section_offset == 0) {
        std::print(std::cerr, "No relocation section found\n");
        return;
    }

    // calculate the RVA of the offset we want to remove
    std::uint32_t target_rva = 0;

    struct TargetInfo {
        std::size_t offset_to_remove;
        std::uint32_t target_rva;
    } target_info = {offset_to_remove, 0};

    auto target_callback = [](void* cbd, const VA& base, const std::string& name, const image_section_header& sec,
                              const bounded_buffer* data) -> int {
        auto* info = static_cast<TargetInfo*>(cbd);

        if (info->offset_to_remove >= sec.PointerToRawData &&
            info->offset_to_remove < sec.PointerToRawData + sec.SizeOfRawData) {
            info->target_rva = sec.VirtualAddress + (info->offset_to_remove - sec.PointerToRawData);
            return 1;
        }

        return 0;
    };

    IterSec(parser_.get(), target_callback, &target_info);
    target_rva = target_info.target_rva;

    if (target_rva == 0) {
        std::print(std::cerr, "Could not find section containing the offset {}\n", offset_to_remove);
        return;
    }

    std::print("Searching for relocation entry for RVA: 0x{:x}\n", target_rva);

    std::size_t reloc_ptr = reloc_section_offset;
    while (reloc_ptr < reloc_section_offset + reloc_section_size) {
        if (reloc_ptr + 8 > buffer_.size()) {
            break;
        }

        std::uint32_t page_rva = *reinterpret_cast<std::uint32_t*>(&buffer_[reloc_ptr]);
        std::uint32_t block_size = *reinterpret_cast<std::uint32_t*>(&buffer_[reloc_ptr + 4]);

        if (block_size == 0 || block_size > reloc_section_size) {
            break;
        }

        // check if our target could be in this block
        if (target_rva >= page_rva && target_rva < page_rva + 0x1000) {
            std::print(
                "Found potential relocation block at offset 0x{:x}, page RVA: "
                "0x{:x}, size: {}\n",
                reloc_ptr, page_rva, block_size);

            std::size_t entries_start = reloc_ptr + 8;
            std::size_t entries_end = reloc_ptr + block_size;

            for (std::size_t entry_ptr = entries_start; entry_ptr < entries_end; entry_ptr += 2) {
                if (entry_ptr + 2 > buffer_.size()) {
                    break;
                }

                std::uint16_t entry = *reinterpret_cast<std::uint16_t*>(&buffer_[entry_ptr]);
                std::uint16_t offset = entry & 0xFFF;
                std::uint32_t entry_rva = page_rva + offset;

                // check if this is our target
                if (entry_rva == target_rva || entry_rva == target_rva + 1 || entry_rva == target_rva + 2 ||
                    entry_rva == target_rva + 3) {
                    std::print("Found relocation entry at offset 0x{:x}, RVA: 0x{:x}\n", entry_ptr, entry_rva);

                    // zero the relocation entry
                    buffer_[entry_ptr] = 0;
                    buffer_[entry_ptr + 1] = 0;

                    std::print("Removed relocation entry\n");
                }
            }
        }

        reloc_ptr += block_size;
    }
}

auto Patcher::patch_checksum_checks() -> void {
    const auto pattern = pattern_t{0x03, 0x06, 0x46, 0x49, 0x75, 0xFA};

    // find "sub [esp+10h+var_10], eax"
    const auto predicate = [](const ZydisInstruction& instr) {
        return instr.backing.info.mnemonic == ZYDIS_MNEMONIC_SUB && instr.backing.info.operand_count >= 2 &&
               instr.backing.operands[0].type == ZYDIS_OPERAND_TYPE_MEMORY &&
               instr.backing.operands[0].mem.base == ZYDIS_REGISTER_ESP &&
               instr.backing.operands[1].reg.value == ZYDIS_REGISTER_EAX;
    };

    for (const auto& offset : Pattern::find_all(buffer_, pattern)) {
        std::print("Checksum loop found at: 0x{:x}\n", image_base_ + offset);

        auto instr = ZydisInstruction::disassemble_until(buffer_, offset, image_base_ + offset, predicate);
        if (!instr) {
            std::print(std::cerr, "Could not find tamper instruction\n");
            continue;
        }

        std::print("Found tamper instruction at: 0x{:x}\n", image_base_ + instr->offset);

        // replace with NOPs
        std::memset(&buffer_[instr->offset], 0x90, 3);
    }
}

auto Patcher::patch_deco_checks() -> void {
    constexpr auto PROGRESSIVE_DECOMPRESS_OFFSET = 20;

    // clang-format off
    const auto pattern = pattern_t{
        0xBA, {}, 0x00, 0x00, 0x00,    // mov edx, trackNumber
        0x52,                          // push edx
        0x33, 0xC0,                    // xor eax, eax
        0xA0, {}, {}, {}, {},          // mov al, driveLetter
        0x50,                          // push eax
    };
    // clang-format on

    const auto cmp_predicate = [](const ZydisInstruction& instr) -> bool {
        // check if instruction is a CMP with immediate value and memory operand
        // using EBP
        return instr.backing.info.mnemonic == ZYDIS_MNEMONIC_CMP && instr.backing.info.operand_count >= 2 &&
               instr.backing.operands[0].type == ZYDIS_OPERAND_TYPE_MEMORY &&
               instr.backing.operands[0].mem.base == ZYDIS_REGISTER_EBP &&
               instr.backing.operands[1].type == ZYDIS_OPERAND_TYPE_IMMEDIATE;
    };

    for (const auto& image_offset : Pattern::find_all(buffer_, pattern)) {
        const std::size_t virtual_address = image_base_ + image_offset;
        const std::size_t drm_address = virtual_address + PROGRESSIVE_DECOMPRESS_OFFSET;

        std::print("Prologue to ProgressiveDecompress_24 found at: 0x{:x}\n", virtual_address);
        std::print("Setup for ProgressiveDecompress_24 at: 0x{:x}\n", drm_address);

        auto cmp_instr = ZydisInstruction::disassemble_until(buffer_, image_offset, virtual_address, cmp_predicate);
        if (!cmp_instr) {
            std::print(std::cerr, "Couldn't find the magic CMP instruction\n");
            continue;
        }

        auto magic_value = static_cast<std::uint32_t>(cmp_instr->backing.operands[1].imm.value.u);
        std::print("Magic value: 0x{:x}\n", magic_value);

        // point to the instruction before the "push edx, ProgressiveDecompress_24"
        const std::size_t patch_offset = image_offset + PROGRESSIVE_DECOMPRESS_OFFSET - 1;

        // ProgressiveDecompress_24 cleans up the stack ("retn 8")
        // so we need to incorporate that into our patch
        // we'll do this by overwriting the following sequence

        // .text:1002B4E6 52              push    edx
        // .text:1002B4E7 BA 76 49 07 10  mov     edx, offset
        // ProgressiveDecompress_24 .text:1002B4EC 52              push    edx
        // .text:1002B4ED C3              retn

        // into the following sequence

        // 023DB4E6 | 83C4 08                  | add esp,8
        // 023DB4E9 | B8 2E0A4B00              | mov eax,MAGIC

        // this also gets rid of the push; ret indirection which is fine

        // 83 C4 08 = ADD ESP, 8
        buffer_[patch_offset + 0] = 0x83;
        buffer_[patch_offset + 1] = 0xC4;
        buffer_[patch_offset + 2] = 0x08;

        // B8 xx xx xx xx = MOV EAX, imm32
        buffer_[patch_offset + 3] = 0xB8;
        buffer_[patch_offset + 4] = static_cast<std::uint8_t>(magic_value & 0xFF);
        buffer_[patch_offset + 5] = static_cast<std::uint8_t>((magic_value >> 8) & 0xFF);
        buffer_[patch_offset + 6] = static_cast<std::uint8_t>((magic_value >> 16) & 0xFF);
        buffer_[patch_offset + 7] = static_cast<std::uint8_t>((magic_value >> 24) & 0xFF);

        std::print("Patched ProgressiveDecompress_24 setup:\n");
        ZydisInstruction::disassemble_until(buffer_, image_offset, virtual_address, cmp_predicate);

        // remove the relocation entry for the ProgressiveDecompress_24 call
        std::print("\n*** Removing relocation entries for ProgressiveDecompress_24 ***\n");
        remove_relocation_entry(patch_offset + 1);  // address bytes start at offset+1
        std::print("\n");
    }
}

auto Patcher::patch_initial_cd_checks() -> void {
    const auto iat_lut = build_iat_lookup_table();

    int offset = 0;
    while (offset < static_cast<int>(buffer_.size())) {
        // we're just bruteforcing through the file, expect errors
        auto instr = ZydisInstruction::disassemble(buffer_, offset, image_base_ + offset);
        if (!instr) {
            offset++;
            continue;
        }

        // skip if not call instruction
        if (instr->backing.info.mnemonic != ZYDIS_MNEMONIC_CALL ||
            instr->backing.operands[0].type != ZYDIS_OPERAND_TYPE_MEMORY) {
            offset++;
            continue;
        }

        // check if this is a call to a CD-check function
        const auto iat_address = instr->backing.operands[0].mem.disp.value;
        if (!iat_lut.contains(iat_address)) {
            offset++;
            continue;
        }

        std::print("Found call to {} at 0x{:x}\n", iat_lut.at(iat_address), image_base_ + offset);

        // find next JCC
        const auto jcc_predicate = [](const ZydisInstruction& instr) {
            return instr.backing.info.mnemonic == ZYDIS_MNEMONIC_JZ ||
                   instr.backing.info.mnemonic == ZYDIS_MNEMONIC_JBE;
        };

        auto jcc_instr = ZydisInstruction::disassemble_until(buffer_, offset, image_base_ + offset, jcc_predicate);
        if (!jcc_instr) {
            offset++;
            continue;
        }

        std::print("Found JCC at 0x{:x}\n", image_base_ + jcc_instr->offset);

        const auto invert_jz = [](std::vector<std::uint8_t>& data, const ZyanUSize offset) {
            if (data[offset] == 0x74) {
                data[offset] = 0x75;  // JNZ short
            } else if (data[offset] == 0x0F && data[offset + 1] == 0x84) {
                data[offset + 1] = 0x85;  // JNZ far
            }
        };

        const auto invert_jbe = [](std::vector<std::uint8_t>& data, const ZyanUSize offset) {
            if (data[offset] == 0x76) {
                data[offset] = 0x77;  // JNBE short
            } else if (data[offset] == 0x0F && data[offset + 1] == 0x86) {
                data[offset + 1] = 0x87;  // JNBE far
            }
        };

        // invert JCC
        switch (jcc_instr->backing.info.mnemonic) {
            case ZYDIS_MNEMONIC_JZ:
                invert_jz(buffer_, jcc_instr->offset);
                break;
            case ZYDIS_MNEMONIC_JBE:
                invert_jbe(buffer_, jcc_instr->offset);
                break;
            default:
                std::print(std::cerr, "Unknown JCC at 0x{:x}\n", image_base_ + jcc_instr->offset);
                return;
        }

        // dump to confirm patch
        ZydisInstruction::disassemble_until(buffer_, offset, image_base_ + offset, [](const ZydisInstruction& instr) {
            return instr.backing.info.mnemonic == ZYDIS_MNEMONIC_JNZ ||
                   instr.backing.info.mnemonic == ZYDIS_MNEMONIC_JNBE;
        });

        return;
    }

    std::print(std::cerr, "Could not find CD TOC check to patch\n");
}

File: src/patcher.rs
use crate::pattern;
use goblin::{Object, pe::PE};
use iced_x86::{Decoder, DecoderOptions, Instruction, Mnemonic, OpKind, Register};
use std::collections::HashMap;

pub struct Patcher {
    buffer: Vec<u8>,
}

impl Patcher {
    pub fn new(buffer: Vec<u8>) -> Self {
        Self { buffer }
    }

    pub fn buffer(&self) -> &[u8] {
        &self.buffer
    }

    pub fn get_base_address(&self) -> usize {
        self.with_pe_image(|pe| {
            pe.header.optional_header.unwrap().windows_fields.image_base as usize
        })
    }

    pub fn patch_checksum_checks(&mut self) {
        stdlog!("\n*** Patching checksum checks ***");

        let pattern = pattern!({0x03, 0x06, 0x46, 0x49, 0x75, 0xFA});
        let results = pattern::find_all(&self.buffer, pattern);
        if results.is_empty() {
            stdlog!("No checksum checks found\n");
            return;
        }

        for idx in results {
            stdlog!(
                "Found checksum check at 0x{:04X}",
                self.get_base_address() + idx
            );
            // find "sub [esp+10h+var_10], eax"
            let instr = Self::disassemble_until(
                &self.buffer[idx..],
                self.get_base_address() + idx,
                |instruction| {
                    instruction.mnemonic() == Mnemonic::Sub
                        && instruction.op_count() == 2
                        && instruction.op0_kind() == OpKind::Memory
                        && instruction.op1_register() == Register::EAX
                },
                false,
            )
            .unwrap_or_else(|| panic!("Failed to find checksum fail instruction at 0x{:04X}", idx));
            stdlog!(
                "Patching checksum fail instruction at 0x{:04X}:",
                instr.ip()
            );

            // NOP out the instruction
            let physical_offset = instr.ip() as usize - self.get_base_address();
            self.buffer[physical_offset..physical_offset + 3].copy_from_slice(&[0x90; 3]);

            // print for comparison
            Self::disassemble_until(
                &self.buffer[idx..],
                self.get_base_address() + idx,
                |instruction| instruction.mnemonic() == Mnemonic::Nop,
                false,
            );

            stdlog!("");
        }
    }

    pub fn patch_early_cd_checks(&mut self) {
        stdlog!("*** Patching early CD checks ***");

        // we need to find the calls used to check for CD drives
        // we found that if the function fails, the early checks are simply skipped
        // without any side effects
        const GET_LOGICAL_DRIVES_LUT: [&str; 2] = ["GetLogicalDrives", "GetLogicalDriveStringsA"];
        let drm_imports = self.with_pe_image(|pe| {
            let mut used_imports = HashMap::new();
            for import in &pe.imports {
                if GET_LOGICAL_DRIVES_LUT.contains(&import.name.as_ref()) {
                    used_imports.insert(import.offset, String::from(import.name.clone()));
                    stdlog!(
                        "Found early CD check function in IAT: {} @ 0x{:04X}",
                        import.name,
                        self.get_base_address() + import.offset
                    );
                }
            }
            used_imports
        });

        if drm_imports.is_empty() {
            stdlog!("No early CD checks found");
            return;
        }

        // find the call instruction that calls the function
        // TODO: we only look for the first hit, what happens if there are multiple?
        // never had that case, but we should be careful
        let call_instr = Self::disassemble_until(
            &self.buffer,
            self.get_base_address(),
            |instruction| {
                instruction.mnemonic() == Mnemonic::Call
                    && instruction.op0_kind() == OpKind::Memory
                    && drm_imports.contains_key(
                        &(instruction.ip_rel_memory_address() as usize)
                            .wrapping_sub(self.get_base_address()),
                    )
            },
            true,
        )
        .unwrap_or_else(|| panic!("Failed to find call instruction for early CD check"));
        stdlog!(
            "Found call instruction for early CD check at 0x{:04X}",
            call_instr.ip()
        );

        // find the next JCC and invert it
        // we may need to find other types of JCC instructions as well
        // i've only encountered JBE and JE so far tho
        let jcc_instr = Self::disassemble_until(
            &self.buffer[call_instr.ip() as usize - self.get_base_address()..],
            call_instr.ip() as usize,
            |instruction| {
                instruction.mnemonic() == Mnemonic::Jbe || instruction.mnemonic() == Mnemonic::Je
            },
            false,
        )
        .unwrap_or_else(|| {
            panic!(
                "Failed to find JCC instruction after call instruction at 0x{:04X}",
                call_instr.ip()
            )
        });
        stdlog!("Found JCC instruction at 0x{:04X}", jcc_instr.ip());

        // invert the JCC instruction
        let physical_offset = jcc_instr.ip() as usize - self.get_base_address();
        match jcc_instr.mnemonic() {
            Mnemonic::Je if self.buffer[physical_offset] == 0x74 => {
                self.buffer[physical_offset] = 0x75 // JNE near
            }
            Mnemonic::Je if self.buffer[physical_offset] == 0x0F => {
                self.buffer[physical_offset + 1] = 0x85 // JNE far
            }
            Mnemonic::Jbe if self.buffer[physical_offset] == 0x76 => {
                self.buffer[physical_offset] = 0x77 // JNBE near
            }
            Mnemonic::Jbe if self.buffer[physical_offset] == 0x0F => {
                self.buffer[physical_offset + 1] = 0x87 // JNBE far
            }
            _ => panic!("Unsupported JCC instruction"),
        };

        stdlog!(
            "\nPatched JCC instruction at 0x{:04X}:",
            jcc_instr.ip() as usize
        );

        // disassemble to see if it worked
        Self::disassemble_until(
            &self.buffer[call_instr.ip() as usize - self.get_base_address()..],
            call_instr.ip() as usize,
            |instruction| instruction.ip() == jcc_instr.ip(),
            false,
        );
    }

    pub fn patch_deco_checks(&mut self) {
        stdlog!("\n*** Patching ProgressiveDecompress_24 CD TOC checks ***");

        let pattern = pattern!({
            0xBA, {}, 0x00, 0x00, 0x00,    // mov edx, trackNumber
            0x52,                          // push edx
            0x33, 0xC0,                    // xor eax, eax
            0xA0, {}, {}, {}, {},          // mov al, driveLetter
            0x50,                          // push eax
        });

        // find the ProgressiveDecompress_24 call prologue
        for idx in pattern::find_all(&self.buffer, pattern) {
            stdlog!(
                "Found pattern for ProgressiveDecompress_24 at 0x{:04X}:",
                self.get_base_address() + idx
            );

            // decode until we find the cmp instruction that verifies the TOC magic value
            let instr = Self::disassemble_until(
                &self.buffer[idx..],
                self.get_base_address() + idx,
                |instruction| {
                    instruction.mnemonic() == Mnemonic::Cmp
                        && instruction.op_count() >= 2
                        && instruction.op0_kind() == OpKind::Memory
                        && instruction.op1_kind() == OpKind::Immediate32
                },
                false,
            )
            .unwrap_or_else(|| {
                panic!(
                    "Failed to find cmp instruction for ProgressiveDecompress_24 at 0x{:04X}",
                    self.get_base_address() + idx
                )
            });
            stdlog!(
                "Prologue to ProgressiveDecompress_24 found at 0x{:04X}",
                instr.ip()
            );

            // we found the TOC magic value
            let magic_value = instr.immediate(1) as u32;
            stdlog!("TOC magic value found: 0x{:04X}", magic_value);

            // set the physical offset to the push just before ProgressiveDecompress_24 gets mov'd
            const PROGRESSIVE_DECOMPRESS_OFFSET: usize = 20;
            let physical_offset = idx + PROGRESSIVE_DECOMPRESS_OFFSET - 1;

            // ProgressiveDecompress_24 cleans up the stack ("retn 8")
            // so we need to incorporate that into our patch
            // we'll do this by overwriting the following sequence

            // .text:1002B4E6 52              push    edx
            // .text:1002B4E7 BA 76 49 07 10  mov     edx, offset
            // .text:1002B4EC 52              push    edx
            // .text:1002B4ED C3              retn

            // into the following sequence

            // 023DB4E6 | 83C4 08                  | add esp,8
            // 023DB4E9 | B8 2E0A4B00              | mov eax,MAGIC

            // this also gets rid of the push; ret indirection which is fine

            // 83 C4 08 = ADD ESP, 8
            self.buffer[physical_offset + 0] = 0x83;
            self.buffer[physical_offset + 1] = 0xC4;
            self.buffer[physical_offset + 2] = 0x08;

            // B8 xx xx xx xx = MOV EAX, imm32
            self.buffer[physical_offset + 3] = 0xB8;
            self.buffer[physical_offset + 4] = ((magic_value >> 0) & 0xFF) as u8;
            self.buffer[physical_offset + 5] = ((magic_value >> 8) & 0xFF) as u8;
            self.buffer[physical_offset + 6] = ((magic_value >> 16) & 0xFF) as u8;
            self.buffer[physical_offset + 7] = ((magic_value >> 24) & 0xFF) as u8;

            stdlog!("\nPatched ProgressiveDecompress_24 call:");
            Self::disassemble_until(
                &self.buffer[physical_offset..],
                self.get_base_address() + physical_offset,
                |instruction| {
                    instruction.mnemonic() == Mnemonic::Cmp
                        && instruction.op_count() >= 2
                        && instruction.op0_kind() == OpKind::Memory
                        && instruction.op1_kind() == OpKind::Immediate32
                },
                false,
            );

            stdlog!("\nRemoving relocation entry at 0x{:04X}", physical_offset);
            self.remove_relocation_entry(physical_offset + 1);
        }
    }

    fn disassemble_until<F>(
        buffer: &[u8],
        start_address: usize,
        predicate: F,
        quiet: bool,
    ) -> Option<Instruction>
    where
        F: Fn(&Instruction) -> bool,
    {
        let mut decoder = Decoder::with_ip(32, buffer, start_address as u64, DecoderOptions::NONE);

        while decoder.can_decode() {
            let instruction = decoder.decode();
            if !quiet {
                stdlog!("0x{:04X}: {}", instruction.ip(), instruction);
            }

            if predicate(&instruction) {
                return Some(instruction);
            }
        }

        None
    }

    fn remove_relocation_entry(&mut self, offset_to_remove: usize) {
        let (reloc_offset, reloc_size) = self.with_pe_image(|pe| {
            pe.sections
                .iter()
                .find(|sec| &sec.name[..6] == b".reloc")
                .map(|sec| {
                    (
                        sec.pointer_to_raw_data as usize,
                        sec.size_of_raw_data as usize,
                    )
                })
                .unwrap_or((0, 0))
        });

        if reloc_offset == 0 || reloc_size == 0 {
            stdlog!("No relocation section found");
            return;
        }

        // find section containing target offset & convert to RVA
        let target_rva = self
            .with_pe_image(|pe| {
                pe.sections
                    .iter()
                    .find(|sec| {
                        let start = sec.pointer_to_raw_data as usize;
                        let end = start + sec.size_of_raw_data as usize;
                        offset_to_remove >= start && offset_to_remove < end
                    })
                    .map(|sec| {
                        let section_offset = offset_to_remove - sec.pointer_to_raw_data as usize;
                        sec.virtual_address as usize + section_offset
                    })
            })
            .expect("Failed to find target section");

        let mut pos = reloc_offset;
        while pos + 8 <= reloc_offset + reloc_size {
            // read page RVA and block size
            let page_rva =
                u32::from_le_bytes(self.buffer[pos..pos + 4].try_into().unwrap()) as usize;
            let block_size =
                u32::from_le_bytes(self.buffer[pos + 4..pos + 8].try_into().unwrap()) as usize;

            if block_size == 0 || block_size > reloc_size {
                break;
            }

            if target_rva >= page_rva && target_rva < page_rva + 0x1000 {
                stdlog!(
                    "Found potential relocation block at offset 0x{:x}, page RVA: 0x{:x}, size: {}",
                    pos,
                    page_rva,
                    block_size
                );

                // process entries in this block
                for entry_pos in (pos + 8..pos + block_size).step_by(2) {
                    if entry_pos + 2 > self.buffer.len() {
                        break;
                    }

                    let entry = u16::from_le_bytes(
                        self.buffer[entry_pos..entry_pos + 2].try_into().unwrap(),
                    ) as usize;
                    let offset = entry & 0xFFF;
                    let entry_rva = page_rva + offset;

                    // check if entry matches target
                    if entry_rva == target_rva
                        || entry_rva == target_rva + 1
                        || entry_rva == target_rva + 2
                        || entry_rva == target_rva + 3
                    {
                        stdlog!(
                            "Found relocation entry at offset 0x{:x}, RVA: 0x{:x}",
                            entry_pos,
                            entry_rva
                        );

                        // zero the entry
                        self.buffer[entry_pos] = 0;
                        self.buffer[entry_pos + 1] = 0;

                        stdlog!("Removed relocation entry");
                    }
                }
            }

            pos += block_size;
        }
    }

    fn get_pe_image(&self) -> PE<'_> {
        match Object::parse(&self.buffer) {
            Ok(Object::PE(pe)) => pe,
            _ => panic!("Corrupted PE buffer"),
        }
    }

    fn with_pe_image<F, R>(&self, f: F) -> R
    where
        F: FnOnce(&PE<'_>) -> R,
    {
        let pe = self.get_pe_image();
        f(&pe)
    }
}
